torch/fx/passes/infra/partitioner.py

```diff
@@ -42,137 +41,104 @@ def __init__(self,
-
         # assumptions: nodes in candidate list is sorted in topological order
         assignment: Dict[Node, int] = {}   # maping from node to partition_id
         partitions_by_id: Dict[int, Partition] = {}  # mapping from partition_id to partition
         new_partition_id = itertools.count()
 
-        def assign(node: Node, id: Optional[int] = None):
-            # If id is None, remove the node from original assigment
-
-            # node has been assigned before, clean up and re-assign
-            if node in assignment:
-                original_id = assignment[node]
-                del assignment[node]
-                partitions_by_id[original_id].remove_node(node)
-                if partitions_by_id[original_id].size() == 0:
-                    del partitions_by_id[original_id]
-
-            if id is not None:
-                assignment[node] = id
-                if id not in partitions_by_id:
-                    partitions_by_id[id] = Partition(id=id, nodes=[node])
+        # try to merge partition other_id into partition self_id
+        # merge only happens if the end graph doesn't contain cyclic dependency
+        # returns `True` when merge happens, `False` otherwise.
+        def maybe_merge_partition(self_id: int, other_id: int):
+            # merged_nodes is the union of nodes in two partition to-be-merged
+            merged_nodes = copy(partitions_by_id[self_id].nodes)
+            merged_nodes.update(partitions_by_id[other_id].nodes)
+
+            # def merge_breaks_dagpartitions: List[Partition]):
+            visited: Set[Node] = set()
+
+            def dfs_find_cycle(node):
+                if node in visited:
+                    return False
+                if node in merged_nodes:
+                    return True  # found cycle, return
+
+                visited.add(node)
+                # branching on hitting partition or not
+                if node in assignment:
+                    # Since partition is not merged in the graph yet, when we
+                    # hit a node in a partition through DFS, we need to
+                    # traverse all nodes in the partition to properly reflect
+                    # dependencies after the fusion
+                    for p_node in partitions_by_id[assignment[node]].nodes:
+                        for user_node in p_node.users:
+                            if dfs_find_cycle(user_node):
+                                return True
                 else:
-                    partitions_by_id[id].add_node(node)
+                    for user_node in node.users:
+                        if dfs_find_cycle(user_node):
+                            return True
+                return False
+
+            # check if merge would create cyclic dependency.
+            for node in merged_nodes:
+                for user_node in node.users:
+                    if user_node not in merged_nodes and dfs_find_cycle(user_node):
```

Comment:
ah... this is a lot of DFS... and duplicated computation... 
is performance a concern here? 

I am wondering if it's better to use a dynamic maintained dependency_map to keep track of this... the look up complexity for dependency_map would be O(1)...

Reply:
> this is a lot of DFS... and duplicated computation...
> I am wondering if it's better to use a dynamic maintained dependency_map to keep track of this

I don't think there is *a lot of* duplicated computation . We are traversing each node strictly once for each attempt to merge two nodes, since we record all `visited`.
If we are to cache a `dependency_map`, when a merge succeeds, we need to traverse and update the `dependency_map` for every node to reflect the updated dependency map.

So between those two approach, I don't think one is obviously faster then the other. Not caching the dependency map definitely looks much simpler on the implementation side and we don't have to constantly deal with modifying a nested table. If I'm to take a guess I tend to think that this would run faster in more cases :laughing:

Reply:
I actually ran into the error like max depth of recursion is reached 🥲

Reply:
Ha... That's.... a pleasant surprise to see that we actually run into big graphs for fusion

I should have some bandwidth tomorrow to refactor this. Should be pretty straightforward. I'll cc you in the PR once I get it to work.
Also if you don't hear from me in a few days, feel free to ping and yell at me to get me back on it :laughing:

Reply:
Here we go: #91042